File: nattybear/level1/신고-결과-받기/Solution.java
import java.util.*;

class Reporters extends HashSet<String> {}

class Graph extends HashMap<String,Reporters> {
}

class Solution {
  public int[] solution(String[] id_list, String[] report, int k) {
    Graph graph = new Graph();

    for (String id : id_list)
      graph.put(id, new Reporters());

    for (String line : report) {
      String[] pair = line.split(" ");
      String id = pair[1];
      String reporter = pair[0];
      graph.get(id).add(reporter);
    }

    return new int[] {};
  }
}

class Counter extends HashMap<String,Integer> {
  public Counter(String[] id_list) {
    super();
    for (String id : id_list)
      super.put(id, 0);
  }

  public void add(String id) {
    int current = super.get(id);
    int new_current = current + 1;
    super.put(id, new_current);
  }
}

File: nattybear/level1/신고-결과-받기/solution.js
class Counter {
  constructor(id_list) {
    this.counter = new Map()
    id_list.map(x => this.counter.set(x, 0))
  }

  add(id) {
    const current = this.counter.get(id)
    const new_current = current + 1
    this.counter.set(id, new_current)
  }

  to_array() {
    return Array.from(this.counter.values())
  }
}

function solution(id_list, report, k) {
  const graph = new Map()
  id_list.map(x => graph.set(x, new Set()))

  report
    .map(x => x.split(' '))
    .map(pair => graph.get(pair[1]).add(pair[0]))

  const counter = new Counter(id_list)

  Array.from(graph.entries())
    .filter(pair => pair[1].size >= k)
    .map(pair => Array.from(pair[1]).map(x => counter.add(x)))

  return counter.to_array()
}
